### PR TITLE
Fix: handle logical CUDA device IDs for GPUStatsMonitor if `CUDA_VISIBLE_DEVICES` set

### PR DESCRIPTION
**Pull request by XuehaiPan**
## What does this PR do?

CUDA device IDs may be different from the real GPU IDs if `CUDA_VISIBLE_DEVICES` set. For example, if I set `CUDA_VISIBLE_DEVICES="3,4,5,6"`, the CUDA device enumeration will be:

  - cuda:0  -->  GPU 3
  - cuda:1  -->  GPU 4
  - cuda:2  -->  GPU 5
  - cuda:3  -->  GPU 6

The previous version of `GPUStatsMonitor` callback will log the wrong devices when users have set a non-trivial `CUDA_VISIBLE_DEVICES` environment variable.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [X] Did you make sure to update the documentation with your changes? (if necessary)
- [X] Did you write any new necessary tests? (not for typos and docs)
- [X] Did you verify new and existing tests pass locally with your changes?
- [X] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [X] Is this pull request ready for review? (if not, please submit in draft mode)
 - [X] Check that all items from **Before submitting** are resolved
 - [X] Make sure the title is self-explanatory and the description concisely explains the PR
 - [X] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/callbacks/test_gpu_stats_monitor.py [hunk 1]
```diff
@@ -116,6 +116,11 @@ def test_gpu_stats_monitor_no_gpu_warning(tmpdir):
 
 
 def test_gpu_stats_monitor_parse_gpu_stats():
```
Comment (awaelchli) [inline tests/callbacks/test_gpu_stats_monitor.py:118]: don't we also want to add a test where we `patch.dict()` the `os.environ` with cuda visible devices set? what do you think?

File: pytorch_lightning/callbacks/gpu_stats_monitor.py [hunk 2]
```diff
@@ -177,12 +184,13 @@ def _to_float(x: str) -> float:
         return stats
 
     @staticmethod
-    def _parse_gpu_stats(gpu_ids: str, stats: List[List[float]], keys: List[Tuple[str, str]]) -> Dict[str, float]:
+    def _parse_gpu_stats(device_ids: List[int], stats: List[List[float]], keys: List[Tuple[str,
+                                                                                           str]]) -> Dict[str, float]:
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:188]: what happened here ^^

Comment (awaelchli) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:188]:
```suggestion
    def _parse_gpu_stats(
        device_ids: List[int], 
        stats: List[List[float]], 
        keys: List[Tuple[str, str]],
    ) -> Dict[str, float]:
```

File: pytorch_lightning/callbacks/gpu_stats_monitor.py [hunk 3]
```diff
@@ -111,7 +111,14 @@ def on_train_start(self, trainer, pl_module) -> None:
...
 
-        self._gpu_ids = ','.join(map(str, trainer.data_parallel_device_ids))
+        self._device_ids = list(trainer.data_parallel_device_ids)
+
+        cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES', default=None)
+        if cuda_visible_devices is None:
+            self._gpu_ids = ','.join(map(str, self._device_ids))
+        else:
+            cuda_visible_devices = cuda_visible_devices.split(',')
+            self._gpu_ids = ','.join(map(cuda_visible_devices.__getitem__, self._device_ids))
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:121]: could we define these in `__init__` and then put a comment for each, so it is clear that device_ids is for the selected ids, and gpu_ids are the unmasked ids.

File: pytorch_lightning/callbacks/gpu_stats_monitor.py [hunk 4]
```diff
@@ -111,7 +111,14 @@ def on_train_start(self, trainer, pl_module) -> None:
                 f' since gpus attribute in Trainer is set to {trainer.gpus}.'
             )
 
-        self._gpu_ids = ','.join(map(str, trainer.data_parallel_device_ids))
+        self._device_ids = list(trainer.data_parallel_device_ids)
+
+        cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES', default=None)
+        if cuda_visible_devices is None:
+            self._gpu_ids = ','.join(map(str, self._device_ids))
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:118]: you could directly define this as the default for cuda_visible_devices, so the mapping code below simplifies and there should be no need for if-else .)

**Review by awaelchli — COMMENTED** [review]
great observation, thanks for sending the fix!

File: pytorch_lightning/callbacks/gpu_stats_monitor.py (hunk 2 above)
Comment (XuehaiPan) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py]: Oh, it was auto formatted by pre-commit.com hooks.

File: tests/callbacks/test_gpu_stats_monitor.py (hunk 1 above)
Comment (XuehaiPan) [reply tests/callbacks/test_gpu_stats_monitor.py:118]: I think it will depends on the GPU resources of the CI test container. If the test case is `CUDA_VISIBLE_DEVICES="0"`, it is a trivial case. But if set `CUDA_VISIBLE_DEVICES="1,0"`, the container should have at least 2 GPUs. Otherwise, `torch.cuda` will raise `RuntimeError('CUDA Error: invalid device ordinal')` at device parsing. I am not sure how to force the CI test container use multi-GPU.

File: pytorch_lightning/callbacks/gpu_stats_monitor.py (hunk 4 above)
Comment (XuehaiPan) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py]: Should I `import cuda` and use `default=range(torch.cuda.device_count())` here? It cannot be assgined as `default=','.join(map(str, self.device_ids))`.

```python
cuda_visible_devices = default = ','.join(map(str, device_ids))
cuda_visible_devices = cuda_visible_devices.split(',')
gpu_ids = ','.join(map(cuda_visible_devices.__getitem__, device_ids))
```

- device_ids = [1, 2, 0] -> gpu_ids = [2, 0, 1]
- device_ids = [1, 2] -> IndexError

File: pytorch_lightning/callbacks/gpu_stats_monitor.py (hunk 3 above)
Comment (XuehaiPan) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py]: These attributes cannot be assigned in `__init__`, because the `Trainner` object have not created at all. They can be assigned in `callback.setup()`. I have added the comments in `setup()` now.

File: pytorch_lightning/callbacks/gpu_stats_monitor.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py]: What I meant is, you can do something like this:

```python
 cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES', ','.join(map(str, self._device_ids)))
 cuda_visible_devices = cuda_visible_devices.split(',')
 gpu_ids = ','.join(map(cuda_visible_devices.__getitem__, self._device_ids))
```

Comment (awaelchli) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py]: Yep, something like this?
```py

cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES', ','.join(range(torch.cuda.device_count()))
 cuda_visible_devices = cuda_visible_devices.split(',')
 gpu_ids = ','.join(map(cuda_visible_devices.__getitem__, self._device_ids))
```

File: pytorch_lightning/callbacks/gpu_stats_monitor.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py]: Yes, I mean initialize them as empty list in the init.

File: pytorch_lightning/callbacks/gpu_stats_monitor.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py]: Also, it's probably best if we keep `gpu_ids` a list internally, like device_ids.

File: pytorch_lightning/callbacks/gpu_stats_monitor.py [hunk 5]
```diff
@@ -154,6 +167,17 @@ def on_train_batch_end(
...
+    @staticmethod
+    def _get_gpu_ids(device_ids: List[int]) -> str:
+        """Get real GPU IDs"""
+        cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES', default=None)
+        if cuda_visible_devices is None:  # `CUDA_VISIBLE_DEVICES` is unset
+            gpu_ids = ','.join(map(str, device_ids))
+        else:
+            cuda_visible_devices = cuda_visible_devices.split(',')
+            gpu_ids = ','.join(map(cuda_visible_devices.__getitem__, device_ids))
+        return gpu_ids
```
Comment (XuehaiPan) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:179]: I prefer to use the if-else switchs (`default=None`) for:

1. It could make an unnecessary module import and an unnecessary function call if set `default=','.join(map(str, range(torch.cuda.device_count())))` when `CUDA_VISIBLE_DEVICES` is set.
2. For testing, `torch.cuda.device_count()` depends on the runtime environment of the CI test container. It could be hard to write unit tests.

File: pytorch_lightning/callbacks/gpu_stats_monitor.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:215]: It should look better by adding this comma

```suggestion
    def _parse_gpu_stats(device_ids: List[int], stats: List[List[float]], keys: List[Tuple[str,
                                                                                           str]],) -> Dict[str, float]:
```

File: tests/callbacks/test_gpu_stats_monitor.py (hunk 1 above)
Comment (awaelchli) [reply tests/callbacks/test_gpu_stats_monitor.py:118]: There is a decorator for that, simply add `@RunIf(min_gpus=2)`

File: pytorch_lightning/callbacks/gpu_stats_monitor.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py]: 1. python only imports once and then caches it. repeated calls to torch.cuda... have no impact
2. `torch.cuda.device_count` is always available, even on a machine without GPUs (returns 0). We can write the test for a GPU machine and just only run it if gpus are available :)

File: tests/callbacks/test_gpu_stats_monitor.py [hunk 6]
```diff
@@ -116,6 +116,40 @@ def test_gpu_stats_monitor_no_gpu_warning(tmpdir):
...
+def test_gpu_stats_monitor_get_gpu_ids_cuda_visible_devices_unset():
+    # `CUDA_VISIBLE_DEVICES` unset
+    try:
+        del os.environ['CUDA_VISIBLE_DEVICES']
+    except KeyError:
+        pass
+
+    gpu_ids = GPUStatsMonitor._get_gpu_ids([1, 0])
+    expected = ['1', '0']
+    assert gpu_ids == expected
```
Comment (XuehaiPan) [inline tests/callbacks/test_gpu_stats_monitor.py:139]: The `CUDA_VISIBLE_DEVICES` should always been unset before the test program starting. The manipulation with `os.environ['CUDA_VISIBLE_DEVICES']` will not work after `torch.cuda` is imported.

```python
$ unset CUDA_VISIBLE_DEIVCES
$ ipython3
In [1]: import torch

In [2]: torch.cuda.device_count()
Out[2]: 10

In [3]: import os

In [4]: os.environ['CUDA_VISIBLE_DEVICES'] = '0'

In [5]: torch.cuda.device_count()
Out[5]: 10
```

```python
$ unset CUDA_VISIBLE_DEIVCES
$ ipython3
In [1]: import os

In [2]: os.environ['CUDA_VISIBLE_DEVICES'] = '0'

In [1]: import torch

In [3]: torch.cuda.device_count()
Out[3]: 1
```

**Comment by tchaton** [comment]
Hey @XuehaiPan,

Mind removing this line there: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L230

It was relying was filtering the available GPUS. There is an associated test with it.

Best,
T.C

File: pytorch_lightning/callbacks/gpu_stats_monitor.py [hunk 7]
```diff
@@ -111,14 +121,20 @@ def on_train_start(self, trainer, pl_module) -> None:
                 f' since gpus attribute in Trainer is set to {trainer.gpus}.'
             )
 
-        self._gpu_ids = ','.join(map(str, trainer.data_parallel_device_ids))
+        # The logical device IDs for selected devices
+        self._device_ids = sorted(set(trainer.data_parallel_device_ids))  # type: List[int]
```
Comment (Borda) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:125]:
```suggestion
        self._device_ids: List[int] = sorted(set(trainer.data_parallel_device_ids))
```

File: pytorch_lightning/callbacks/gpu_stats_monitor.py [hunk 8]
```diff
@@ -111,14 +121,20 @@ def on_train_start(self, trainer, pl_module) -> None:
...
+
+        # The unmasked real GPU IDs
+        self._gpu_ids = self._get_gpu_ids(self._device_ids)  # type: List[str]
```
Comment (Borda) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:128]:
```suggestion
        self._gpu_ids: List[int] = self._get_gpu_ids(self._device_ids)
```

File: pytorch_lightning/callbacks/gpu_stats_monitor.py [hunk 9]
```diff
@@ -147,19 +169,30 @@ def on_train_batch_end(
...
 
         trainer.logger.log_metrics(logs, step=trainer.global_step)
 
+    @staticmethod
+    def _get_gpu_ids(device_ids: List[int]) -> List[str]:
+        """Get the unmasked real GPU IDs"""
+        cuda_visible_devices = os.getenv(
+            'CUDA_VISIBLE_DEVICES',
+            default=','.join(map(str, range(torch.cuda.device_count())))  # all devices if `CUDA_VISIBLE_DEVICES` unset
+        ).split(',')
```
Comment (Borda) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:185]:
```suggestion
        # all devices if `CUDA_VISIBLE_DEVICES` unset
        cuda_visible_devices = os.getenv(
            'CUDA_VISIBLE_DEVICES',
            default=','.join(map(str, range(torch.cuda.device_count())))
        ).split(',')
```

File: tests/callbacks/test_gpu_stats_monitor.py [hunk 10]
```diff
@@ -116,6 +116,40 @@ def test_gpu_stats_monitor_no_gpu_warning(tmpdir):
...
+    expected = {
+        'device_id: 1/gpu (a)': 3,
+        'device_id: 1/memory (b)': 4,
+        'device_id: 2/gpu (a)': 6,
+        'device_id: 2/memory (b)': 7
+    }
     assert logs == expected
+
+
+@RunIf(min_gpus=2)
```
Comment (Borda) [inline tests/callbacks/test_gpu_stats_monitor.py:129]: rather mock the environ

File: tests/callbacks/test_gpu_stats_monitor.py [hunk 11]
```diff
@@ -116,6 +116,40 @@ def test_gpu_stats_monitor_no_gpu_warning(tmpdir):
...
+        del os.environ['CUDA_VISIBLE_DEVICES']
+    except KeyError:
+        pass
+
+    gpu_ids = GPUStatsMonitor._get_gpu_ids([1, 0])
+    expected = ['1', '0']
+    assert gpu_ids == expected
+
+
+def test_gpu_stats_monitor_get_gpu_ids_cuda_visible_devices_set():
```
Comment (Borda) [inline tests/callbacks/test_gpu_stats_monitor.py:142]: again mock environ

**Comment by XuehaiPan** [comment]
> Mind removing this line there: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L230

`gpus_metrics` contains all GPUs' `memory.used` fields, but they are enumerated in the real GPU IDs. If `CUDA_VISIBLE_DEVICES` is set to integer strings, it's easy to map the GPU IDs to CUDA device IDS. But if `CUDA_VISIBLE_DEVICES` is set to UUID strings, it could be hard to determine the device mapping.

File: tests/callbacks/test_gpu_stats_monitor.py [hunk 12]
```diff
@@ -116,6 +117,39 @@ def test_gpu_stats_monitor_no_gpu_warning(tmpdir):
...
+@mock.patch.dict(os.environ, {})
+@mock.patch('torch.cuda.device_count', return_value=2)
+@mock.patch('torch.cuda.is_available', return_value=True)
+def test_gpu_stats_monitor_get_gpu_ids_cuda_visible_devices_unset(device_count_mock, is_avaliable_mock):
+    gpu_ids = GPUStatsMonitor._get_gpu_ids([1, 0])
+    expected = ['1', '0']
+    assert gpu_ids == expected
+
+
+@mock.patch.dict(os.environ, {'CUDA_VISIBLE_DEVICES': '2,3,4'})
```
Comment (awaelchli) [inline tests/callbacks/test_gpu_stats_monitor.py:140]: nit: we could permute the numbers here :)

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
Pushed a commit to use list comprehensions instead which are simpler to understand

**Review by justusschock — APPROVED** [review]
